feat: bidirectional support

jnumainville opened the pull request:
Fixes #2 

Adds python support for bidirectional communication and updating of plots with plot bys. 
Tested with Matt's front-end changes (and will need to be merged with those). 
Simple example plot:
```
import deephaven.plot.express as dx
from deephaven import time_table
import random 

sourceh = time_table("PT2S").update(formulas=[
    "X = (int)random.gauss(20, 5)", 
    "Z = (float)random.gauss(30, 3)",
    "C = i % 20"
    ])

fig = dx.scatter(sourceh, x="X", y="Z",
    by="C",
)
```

Also has some miscellaneous cleanup and refactoring.

jnumainville commented on plugins/plotly-express/src/deephaven/plot/express/communication/DeephavenFigureListener.py:
```diff
@@ -0,0 +1,167 @@
+        """
+        # only one export can be done at a time to ensure that the references
+        # are correct
+        exporter.acquire()
```
Am going to move this to a context manager. Although I wonder if I have a correct understanding that a TableListener can lead to a race condition. My understanding on how the java and python interact here is weak.

mattrunyon replied:
This lock may have been doing something else that is necessary. I'm getting fairly frequent deadlocks when plotting w/ the latest change. Usually on re-creating a figure, but have gotten it once on the first time I plotted something after a fresh server start

mattrunyon replied:
Nevermind I was able to deadlock even building off the commit before the lock was removed

jnumainville replied:
I'm seeing those deadlocks too. I'm not sure where they're coming from. It seems to be the whole python interpreter that gets locked.

jnumainville replied:
@mattrunyon It might have been some figure locks I had that shouldn't also be needed. I removed them and there doesn't seem to be deadlocks any more.
I am still seeing odd behavior when switching between figure tabs. Is there any possibility when switching on and off a figure that there is some lingering data in the window that could mess up the new figure creation? As far as I can tell the proper updates are being sent.

mattrunyon replied:
Pushed a fix for that stale data issue. Also fixed a double loader that appeared from some CSS changes to web-client-ui.

We'll need to update to non-beta packages on the next release to get TS fixes in for this. If there's no other issues w/ this PR though, could just ts-ignore the TS issue and fix it in a future PR when changing to a `WidgetPlugin`

mattrunyon commented:
Added JS bidirectional support. Looks like I need to rebase on main though.

There are some issues w/ recreating a plot w/ the same name (i.e. run the same code twice in a row in the repl). There is an error printed to the user console because something is double closing I think and causing the error (no way to detect it's already been closed). This also seems to be an issue w/ our tables and charts, but it's a JS error/warning.

Turns out in `Pending.cancel` we call the cleanup on all promises, not just the pending promises. That causes the double close on unmount for everything I think.

jnumainville commented on plugins/plotly-express/src/deephaven/plot/express/communication/DeephavenFigureListener.py:
```diff
@@ -0,0 +1,134 @@
+            connection: MessageStream: The connection to send messages to
+            liveness_scope: The liveness scope to use for the listeners
+        """
+        self._connection: MessageStream = connection
```
removed MessageStream type

jnumainville commented on package-lock.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "deephaven-js-plugins",
-  "lockfileVersion": 3,
+  "lockfileVersion": 2,
```
fixed I think?

jnumainville commented on plugins/plotly-express/src/deephaven/plot/express/communication/DeephavenFigureConnection.py:
```diff
@@ -0,0 +1,39 @@
+    def __init__(self, figure: DeephavenFigure, client_connection: MessageStream):
+        super().__init__()
+        JLivenessScope = jpy.get_type("io.deephaven.engine.liveness.LivenessScope")
+        self._liveness_scope = JLivenessScope()
```
done

jnumainville commented:
@mattrunyon it seems like when I switch panels (from looking at one figure to not then back to looking at it) a new connection is set up but the old one is still there. Do you know how to stop that?

jnumainville commented:
Everything seems to be working properly after Matt's fixes

jnumainville commented:
After discussion with Colin, the locks are something should be done, so I have added them back in. 
The locks in the figure DAG (DeephavenLayerNode and DeephavenFigureNode) were causing a deadlock because some table operations in the figure regeneration needed to lock but couldn't because of the lock in the node. 
I have worked around this by not locking the figure generation itself, only the update of the figure in the node, with a revision number to ensure that an older update doesn't overwrite a newer update.
I wrote a RevisionManager to deal with this issue, so I also converted the exporter-related locks to use the RevisionManager.

jnumainville commented:
@mattrunyon I have (hopefully correctly) added the tables to the liveness scope
I also did something I should have already had, where each session/listener makes a copy of the deephaven figure and the listener updates its specific figure. I think I removed I removed it because it seemed like overkill, but if you're going to have session specific liveness scopes that need to keep track of tables it makes sense anyways. This has the perhaps useful in the future benefit of making it much easier to make session specific updates to the figure, like one_clicks, if we ever wanted to implement those directly into the figure.
Essentially, the session liveness scope is carried along for the figure copy, adding any tables that are in the figure DAG. These are removed on del, either of the base figure or the copy (as well as the session, as before).

jnumainville commented:
@mattrunyon fixed #2 with an empty default figure in the partition case.
The only weird thing is that I have to click and drag the figures to get them to start working

code I used:
```
import deephaven.plot.express as dx
from deephaven import time_table
import random 

sourceh = time_table("PT1S").update(formulas=[
    "X = (int)random.gauss(20, 5)", 
    "Y = (int)random.gauss(20, 5)", 
    "Z = (float)random.gauss(30, 3)",
    "l1 = i % 20",
    "l2 = i % 30",
    ]).tail(5)

fig1 = dx.scatter(sourceh, x="Z", y="Y",
    by="l1",
    color_discrete_sequence=["lemonchiffon"]
)

fig2 = dx.scatter(sourceh, x="Z", y="Y",
    by="l2",
    color_discrete_sequence=["salmon"]
)

layered = dx.layer(fig1, fig2)
```

mofojed commented on plugins/plotly-express/src/js/src/PlotlyExpressChartModel.ts:
```diff
@@ -4,177 +4,310 @@ import type {
-      this.chartDataMap.set(table, new dh.plot.ChartData(table))
+    this.isSubscribed = true;
+    this.widgetUnsubscribe = this.widget.addEventListener(
+      'message',
```
Use `this.dh.Widget.EVENT_MESSAGE`

mofojed commented on plugins/plotly-express/src/js/src/PlotlyExpressChartModel.ts:
```diff
@@ -4,177 +4,310 @@ import type {
+
+    this.tableReferenceMap.forEach((_, id) => this.removeTable(id));
+
+    // TODO: Add this back when JsWidget doesn't log a Python error when already closed
+    // This issue occurs when you recreate a plot in the REPL as the engine closes its side
+    // but the widget doesn't know and unsubscribe on panel unmount causes this close
+    // this.widget?.close();
```
Is there a ticket already for this? I know Colin said he was going to fix it, but did it slide off his radar?

mattrunyon replied:
https://github.com/deephaven/deephaven-core/issues/4604

I'm going to remove this comment and uncomment the code. Either way there's errors printed to the user's console, but I think once that bug is fixed this will be the correct way to do it

jnumainville commented on plugins/plotly-express/src/deephaven/plot/express/communication/DeephavenFigureListener.py:
```diff
@@ -0,0 +1,166 @@
+        """
+        # need to create a new exporter for each message
+        message = json.loads(payload.decode())
+        if message["type"] == "RETRIEVE":
```
created #109

jnumainville commented on plugins/plotly-express/src/deephaven/plot/express/plots/_private_utils.py:
```diff
@@ -258,93 +340,132 @@ def sync_pop(self):
+    set_all(args, BOX_DEFAULTS)
+
+    func = px.box
+    groups = {"marker", "preprocess_violin", "supports_lists"}
```
Yes, violin, box, and strip all have some overlapping preprocessing steps as they have many shared steps. I named this group preprocess_violin originally because I couldn't think of a good category name for that grouping. I'll switch it to `preprocess_spread`. I'll also pull the groups in a constant.

mattrunyon commented on plugins/plotly-express/src/js/src/PlotlyExpressChartModel.ts:
```diff
@@ -4,177 +4,310 @@ import type {
+    this.tableColumnReplacementMap.forEach((columnReplacements, tableId) => {
+      const tableData = this.tableDataMap.get(tableId);
+      if (tableData == null) {
+        // log.warn('No tableData for this event. Skipping update');
```
I changed it to throw. It _shouldn't_ ever happen, but if it does then there's probably some out of sync state or stale state laying around

Can change back to a warning if you think that's right. It doesn't cause an issue to ignore it, but the plot might not be correct if this case were to happen. Prefer error over showing potentially incorrect data?

mofojed replied:
Yup error over showing incorrect.

mattrunyon replied:
I need to fix something for initial loading actually with this. The chart is finished loading after the model initializes, but before the tables are fetched which leads to this throwing

mattrunyon replied:
Ok I have changed this to throw and set an empty tableData when the table is added. There could be states (like initial load) where the table is being fetched and doesn't have its data yet, so that's why I added the default empty tableData

The error should only throw if something bad happens or somehow gets out of sync. Mostly there so TS doesn't complain about the possibility of null

jnumainville commented:
At this point I have resolved all bugs and comments I am aware of in the python side of the application

mattrunyon commented:
I believe the JS side has any known issues fixed as well
